test/e2e/mtls/standalone_universal.go

```diff
@@ -0,0 +1,139 @@
+		Expect(universal.VerifyKuma()).To(Succeed())
+	})
+
+	E2EAfterEach(func() {
+		Expect(universal.DeleteKuma(universalOpts...)).To(Succeed())
+		Expect(universal.DismissCluster()).To(Succeed())
+	})
+
+	createMesh := func(name, mode string) {
+		meshYaml := fmt.Sprintf(
+			`
+type: Mesh
+name: %s
+mtls:
+  enabledBackend: ca-1
+  backends:
+  - name: ca-1
+    type: builtin
+    mode: %s`, name, mode)
+		Expect(YamlUniversal(meshYaml)(universal)).To(Succeed())
+	}
+
+	runDemoClient := func(mesh string) {
+		demoClientToken, err := universal.GetKuma().GenerateDpToken(mesh, "demo-client")
+		Expect(err).ToNot(HaveOccurred())
+		Expect(
+			DemoClientUniversal(AppModeDemoClient, mesh, demoClientToken, WithTransparentProxy(true))(universal),
+		).To(Succeed())
+	}
+
+	runTestServer := func(mesh string) {
+		echoServerToken, err := universal.GetKuma().GenerateDpToken(mesh, "test-server")
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(TestServerUniversal("test-server", mesh, echoServerToken,
+			WithArgs([]string{"echo", "--instance", "universal-1"}))(universal)).To(Succeed())
+	}
+
+	checkInsideMeshConnection := func() {
+		Eventually(func() bool {
+			stdout, _, err := universal.Exec("", "", "demo-client", "curl", "-v", "-m", "8", "--fail", "test-server.mesh")
+			if err != nil {
+				return false
+			}
+			return strings.Contains(stdout, "HTTP/1.1 200 OK")
+		}, "30s", "1s").Should(BeTrue())
+	}
+
+	getServiceEndpoint := func() string {
+		var addr string
+		r := regexp.MustCompile(`::([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}):?([0-9]{1,5})?::`)
+		Eventually(func() bool {
+			cmd := []string{"/bin/bash", "-c", "\"curl -s localhost:30001/clusters | grep test-server.*cx_active\""}
+			stdout, _, err := universal.Exec("", "", "demo-client", cmd...)
+			if err != nil {
+				return false
+			}
+			submatch := r.FindStringSubmatch(stdout)
+			if len(submatch) < 3 {
+				return false
+			}
+			addr = fmt.Sprintf("%s:%s", submatch[1], submatch[2])
+			return true
+		}, "30s", "1s").Should(BeTrue())
+		return addr
+	}
+
+	checkOutsideMeshConnection := func() {
+		addr := getServiceEndpoint()
+
+		Eventually(func() bool {
+			stdout, _, err := universal.Exec("", "", "demo-client", "curl", "-v", "-m", "8", "--fail", addr)
+			if err != nil {
+				return false
+			}
+			return strings.Contains(stdout, "HTTP/1.1 200 OK")
+		}, "30s", "1s").Should(BeTrue())
+	}
+
+	checkNoOutsideMeshConnection := func() {
+		addr := getServiceEndpoint()
+
+		Eventually(func() bool {
+			stdout, _, err := universal.Exec("", "", "demo-client", "curl", "-v", "-m", "8", "--fail", addr)
+			if err != nil {
+				return false
+			}
+			return strings.Contains(stdout, "HTTP/1.1 200 OK")
+		}, "30s", "1s").Should(BeFalse())
+	}
+
+	It("should support STRICT mTLS mode", func() {
+		createMesh("default", "STRICT")
+
+		runTestServer("default")
+
+		runDemoClient("default")
+
+		checkInsideMeshConnection()
+
+		checkNoOutsideMeshConnection()
```

Comment:
Disclaimer: My opinion, your opinion may vary.

Please try to not overuse functions inside a test. For me, it's easier to read a test with code embedded
```
// given a mesh
... code that creates a mesh

// and test server
... code that creates test server
```

than test with functions. To read the full test I need to jump back and forth between test and function implementations.

Of course, if we have reusable functions and repeating the complex operations, it may make sense to use it.

Also, I'd try to keep the given, when, then convention. For me, it makes it easier to read the intention of the test.

Reply:
I agree that it was probably too much. I got rid of these `checkInsideMeshConnection` and `checkNoOutsideMeshConnection`, but I left the functions that create a mesh and runs client, they are 100 percent reusable across tests in this file